### PR TITLE
tests.nixpkgs-check-by-name: Fix for symlinked tempdirs

### PR DESCRIPTION
**Pull request by infinisil**
## Description of changes

Reported by @Atemu at NixCon :). 

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/test/nixpkgs-check-by-name/src/eval.rs [hunk 1]
```diff
@@ -59,7 +59,7 @@ pub fn check_values<W: io::Write>(
         .args(["--arg", "attrsPath"])
         .arg(attrs_file.path())
```
Comment (Atemu) [inline pkgs/test/nixpkgs-check-by-name/src/eval.rs:60]: This one needs to be canonicalised aswell.

```suggestion
        .arg(attrs_file.path().canonicalize()?)
```

Once you do that, it appears to work on Darwin.

Perhaps it'd be better to just do that up top in the `attrs_file` declaration though?

**Review by Atemu — CHANGES_REQUESTED** [review]

**Comment by Atemu** [comment]
For context, what this PR intends to fix is a bug discovered on Darwin where the generated attrs list is put into a location that is behind a symlink. Including that location in the NIX_PATH and then trying to access it results in an error like this:

```
error:
       … while calling anonymous lambda

         at «string»:9:1:

            8| # - is_derivation: The result of `lib.isDerivation <attr>`
            9| {
             | ^
           10|   attrsPath,

       error: access to canonical path '/private/var/folders/xp/9_ry6h9x6l9gh_g32qspz0_40000gp/T/.tmpFbcNO0' is forbidden in restricted mode
```

On Darwin, `/tmp` is a symlink to `/private/tmp/` and the attrs file is put in `/tmp` which means this bug is triggered every time. It should be reproducible on other platforms aswell though.

This might also be a Nix bug.

File: pkgs/test/nixpkgs-check-by-name/src/eval.rs (hunk 1 above)
Comment (infinisil) [reply pkgs/test/nixpkgs-check-by-name/src/eval.rs]: Good idea, done now!

**Comment by infinisil** [comment]
Thanks, I added the error message to the code and commit message. I also added a test to make sure we don't run into this again.

**Review by Atemu — CHANGES_REQUESTED** [review]
With my limited rust knowledge the diff LGTM but at runtime I get:

```
Executing cargoCheckHook
++ cargo test -j 10 --release --target aarch64-apple-darwin --frozen -- --test-threads=10
   Compiling nixpkgs-check-by-name v0.1.0 (/private/tmp/nix-build-nixpkgs-check-by-name.drv-0/source)
    Finished release [optimized] target(s) in 0.79s
     Running unittests src/main.rs (target/aarch64-apple-darwin/release/deps/nixpkgs_check_by_name-c15b59f9dc020547)

running 3 tests
test tests::test_case_sensitive ... ok(B
error: creating symlink from '/private/tmp/nix-build-nixpkgs-check-by-name.drv-0/source/test-tmp/var/nix/gcroots/profiles' to '/private/tmp/nix-build-nixpkgs-check-by-name.drv-0/source/test-tmp/var/nix/profiles': File exists
test tests::tests_dir ... FAILED(B
test tests::test_symlinked_tmpdir ... ok(B

failures:

---- tests::tests_dir stdout ----
Error: Failed test case override-different-file

Caused by:
    Failed to run command "nix-instantiate" "--eval" "--json" "--strict" "--readonly-mode" "--restrict-eval" "--show-trace" "--expr" "# Takes a path to nixpkgs and a path to the json-encoded list of attributes to check.\n# Returns an attribute set containing information on each requested attribute.\n# If the attribute is missing from Nixpkgs it\'s also missing from the result.\n#\n# The returned information is an attribute set with:\n# - call_package_path: The <path> from `<attr> = callPackage <path> { ... }`,\n#   or null if it\'s not defined as with callPackage, or if the <path> is not a path\n# - is_derivation: The result of `lib.isDerivation <attr>`\n{\n  attrsPath,\n  nixpkgsPath,\n}:\nlet\n  attrs = builtins.fromJSON (builtins.readFile attrsPath);\n\n  # This overlay mocks callPackage to persist the path of the first argument\n  callPackageOverlay = self: super: {\n    callPackage = fn: args:\n      let\n        result = super.callPackage fn args;\n      in\n      if builtins.isAttrs result then\n        # If this was the last overlay to be applied, we could just only return the `_callPackagePath`,\n        # but that\'s not the case because stdenv has another overlays on top of user-provided ones.\n        # So to not break the stdenv build we need to return the mostly proper result here\n        result // {\n          _callPackagePath = fn;\n        }\n      else\n        # It\'s very rare that callPackage doesn\'t return an attribute set, but it can occur.\n        {\n          _callPackagePath = fn;\n        };\n  };\n\n  pkgs = import nixpkgsPath {\n    # Don\'t let the users home directory influence this result\n    config = { };\n    overlays = [ callPackageOverlay ];\n  };\n\n  attrInfo = attr: {\n    # These names are used by the deserializer on the Rust side\n    call_package_path =\n      if pkgs.${attr} ? _callPackagePath && builtins.isPath pkgs.${attr}._callPackagePath then\n        toString pkgs.${attr}._callPackagePath\n      else\n        null;\n    is_derivation = pkgs.lib.isDerivation pkgs.${attr};\n  };\n\n  attrInfos = builtins.listToAttrs (map (name: {\n    inherit name;\n    value = attrInfo name;\n  }) attrs);\n\nin\n# Filter out attributes not in Nixpkgs\nbuiltins.intersectAttrs pkgs attrInfos\n" "--arg" "attrsPath" "/private/tmp/nix-build-nixpkgs-check-by-name.drv-0/.tmpsmDyqP/actual/.tmpMJuBGU" "-I" "/private/tmp/nix-build-nixpkgs-check-by-name.drv-0/.tmpsmDyqP/actual/.tmpMJuBGU" "--arg" "nixpkgsPath" "/private/tmp/nix-build-nixpkgs-check-by-name.drv-0/source/tests/override-different-file" "-I" "/private/tmp/nix-build-nixpkgs-check-by-name.drv-0/source/tests/override-different-file" "-I" "tests/mock-nixpkgs.nix"


failures:
    tests::tests_dir

test result: FAILED(B. 2 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.37s

error: test failed, to rerun pass `--bin nixpkgs-check-by-name`
error: builder for '/nix/store/krq0s9qg8ll4rzy46ih1mfbqj0lg2p8s-nixpkgs-check-by-name.drv' failed with exit code 101;
```

I assume this doesn't repro on Linux?

**Comment by infinisil** [comment]
Interesting, this looks a bit like https://github.com/NixOS/nix/issues/2706, though I don't fully understand it yet.

**Comment by infinisil** [comment]
@ofborg build tests.nixpkgs-check-by-name

**Comment by infinisil** [comment]
> Interesting, this looks a bit like [NixOS/nix#2706](https://github.com/NixOS/nix/issues/2706), though I don't fully understand it yet.

Indeed this was it. Apparently cargo runs tests in parallel by default, which triggers this. One way to fix this would be to make cargo run tests sequentially, which could be done with `dontUseCargoParallelTests = true;`, but that's a bit hacky. Instead I'm going to copy the fix for this from https://github.com/tweag/lorri/pull/58.

Btw I'm not sure why this wasn't a problem before when we already had two separate tests, we have three now, oh well :shrug:

**Comment by infinisil** [comment]
@ofborg build tests.nixpkgs-check-by-name

**Comment by infinisil** [comment]
There are some ofborg failures but those are unrelated, they reference something that was removed a while ago, see https://github.com/NixOS/ofborg/pull/650

**Review by Atemu — APPROVED** [review]
Result of `nixpkgs-review pr 254435` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tests.nixpkgs-check-by-name</li>
  </ul>
</details>

```
$ nixpkgs-check-by-name nixpkgs/
Validated successfully
```

**Pull request merged**